DRILL-7716: Create Format Plugin for SPSS Files

cgivre opened the pull request:
# [DRILL-7716](https://issues.apache.org/jira/browse/DRILL-7716): Create Format Plugin for SPSS Files

## Description

This PR adds the ability for Drill to query SPSS files.

## Documentation
# Format Plugin for SPSS (SAV) Files
This format plugin enables Apache Drill to read and query Statistical Package for the Social Sciences (SPSS) (or Statistical Product and Service Solutions) data files. According
 to Wikipedia: [1]
 
***
 SPSS is a widely used program for statistical analysis in social science. It is also used by market researchers, health researchers, survey companies, government, education researchers, marketing organizations, data miners, and others. The original SPSS manual (Nie, Bent & Hull, 1970) has been described as one of "sociology's most influential books" for allowing ordinary researchers to do their own statistical analysis. In addition to statistical analysis, data management (case selection, file reshaping, creating derived data) and data documentation (a metadata dictionary is stored in the datafile) are features of the base software.
***
 
## Configuration 
To configure Drill to read SPSS files, simply add the following code to the formats section of your file-based storage plugin.  This should happen automatically for the default
 `cp`, `dfs`, and `S3` storage plugins.
 
 Other than the file extensions, there are no variables to configure.
 
```json
"spss": {
          "type": "spss",
          "extensions": [
            "sav"
          ]
        }
```

## Data Model
SPSS only supports two data types: Numeric and Strings.  Drill maps these to `DOUBLE` and `VARCHAR` respectively. However, for some numeric columns, SPSS maps these numbers to
 text, similar to an `enum` field in Java.
 
 For instance, a field called `Survey` might have labels as shown below:
 
 <table>
    <tr>
        <th>Value</th>
        <th>Text</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Yes</td>
    </tr>
    <tr>
        <td>2</td>
        <td>No</td>
    </tr>
    <tr>
        <td>99</td>
        <td>No Answer</td>
    </tr>
 </table>

For situations like this, Drill will create two columns. In the example above you would get a column called `Survey` which has the numeric value (1,2 or 99) as well as a column
 called `Survey_value` which will map the integer to the appropriate value. Thus, the results would look something like this:
 
 <table>
 <tr>
 <th>`Survey`</th>
 <th>`Survey_value`</th>
 </tr>
 <tr>
 <td>1</td>
 <td>Yes</td>
 </tr>
  <tr>
  <td>1</td>
  <td>Yes</td>
  </tr>
   <tr>
   <td>1</td>
   <td>Yes</td>
   </tr>
    <tr>
    <td>2</td>
    <td>No</td>
    </tr>
     <tr>
     <td>1</td>
     <td>Yes</td>
     </tr>
      <tr>
      <td>2</td>
      <td>No</td>
      </tr>
  <tr>
  <td>99</td>
  <td>No Answer</td>
  </tr>
 </table>


[1]: https://en.wikipedia.org/wiki/SPSS
## Testing
There are unit tests attached to this PR. 

This PR also includes DRILL-7600 (https://issues.apache.org/jira/browse/DRILL-7600)

vvysotskyi commented on contrib/format-spss/README.md:
```diff
@@ -0,0 +1,87 @@
+This format plugin enables Apache Drill to read and query Statistical Package for the Social Sciences (SPSS) (or Statistical Product and Service Solutions) data files. According
+ to Wikipedia: [1]
+ 
+ SPSS is a widely used program for statistical analysis in social science. It is also used by market researchers, health researchers, survey companies, government, education researchers, marketing organizations, data miners, and others. The original SPSS manual (Nie, Bent & Hull, 1970) has been described as one of "sociology's most influential books" for allowing ordinary researchers to do their  ...
```
Blockquote may be used for quoting from Wiki.

cgivre replied:
Fixed.

vvysotskyi commented on contrib/format-spss/README.md:
````diff
@@ -0,0 +1,87 @@
+ Other than the file extensions, there are no variables to configure.
+ 
+```json
+"spss": {
````
Please fix the formatting.

cgivre replied:
Fixed (I think)

vvysotskyi commented on contrib/format-spss/README.md:
```diff
@@ -0,0 +1,87 @@
+# Format Plugin for SPSS (SAV) Files
+This format plugin enables Apache Drill to read and query Statistical Package for the Social Sciences (SPSS) (or Statistical Product and Service Solutions) data files. According
+ to Wikipedia: [1]
```
Wikipedia may be made as a reference instead of putting the long at the end of the doc.

cgivre replied:
Fixed

vvysotskyi commented on contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java:
```diff
@@ -0,0 +1,242 @@
+
+  private static final String VALUE_LABEL = "_value";
+
+  private final SpssReaderConfig readerConfig;
```
This field is newer used.

cgivre replied:
Removed

vvysotskyi commented on contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java:
```diff
@@ -0,0 +1,242 @@
+
+  private boolean processNextRow() {
+    try {
+      SpssColumnWriter currentColumn;
```
Looks like using this variable may be omitted.

cgivre replied:
Removed

vvysotskyi commented on contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java:
```diff
@@ -0,0 +1,242 @@
+      SpssColumnWriter currentColumn;
+
+      // Stop reading when you run out of data
+      if (!spssReader.readNextCase() ) {
```
```suggestion
      if (!spssReader.readNextCase()) {
```

cgivre replied:
Fixed

vvysotskyi commented on contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java:
```diff
@@ -0,0 +1,242 @@
+    SchemaBuilder builder = new SchemaBuilder();
+    variableList = spssReader.getVariables();
+
+    for(SpssVariable variable : variableList) {
```
```suggestion
    for (SpssVariable variable : variableList) {
```

cgivre replied:
Fixed

vvysotskyi commented on contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java:
```diff
@@ -0,0 +1,242 @@
+
+    ScalarWriter writer;
+
+    public boolean isNumeric;
```
This field is newer accessed, so can we remove it?

cgivre replied:
Removed.

vvysotskyi commented on contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java:
```diff
@@ -0,0 +1,242 @@
+  }
+
+  public abstract static class SpssColumnWriter {
+    String columnName;
```
You may introduce a constructor which accepts these fields and make them final.

cgivre replied:
Added constructor.

vvysotskyi commented on contrib/format-spss/src/main/resources/bootstrap-format-plugins.json:
```diff
@@ -0,0 +1,37 @@
+{
+  "storage":{
```
```suggestion
  "storage": {
```

cgivre replied:
Fixed

vvysotskyi commented on exec/java-exec/src/test/java/org/apache/drill/test/ClusterTest.java:
```diff
@@ -125,4 +139,29 @@ public static void run(String query, Object... args) throws Exception {
+   * @param outFileName the output file name
+   * @throws IOException Throws IO exception if the file cannot be found or any other IO error
+   */
+  public void generateCompressedFile(String fileName, String codecName, String outFileName) throws IOException {
```
Please make it static and move to a more suitable place, for example to `QueryTestUtil`.

cgivre replied:
Moved to `QueryTestUtil`

cgivre replied:
Note:  By moving this class, this PR also addresses Drill-7600: https://issues.apache.org/jira/browse/DRILL-7600.

vvysotskyi commented on protocol/src/main/protobuf/UserBitShared.proto:
```diff
@@ -379,7 +379,11 @@ enum CoreOperatorType {
   SHP_SUB_SCAN = 65;
   METADATA_HANDLER = 66;
   METADATA_CONTROLLER = 67;
+  // 68 Reserved for Apache Druid
```
Please remove these comments. When plugins are merged, new values will be introduced. Protobufs will be regenerated in both cases.

cgivre replied:
Comments removed and protobufs rebuilt.

cgivre commented:
@vvysotskyi 
Thanks for the quick review!  I addressed all your review comments.  Regarding updating the website, I will be happy to do so, but I've never done it before.  I'll create a JIRA to do so and add this to it.

vvysotskyi commented on pom.xml:
```diff
@@ -359,6 +359,7 @@
             <exclude>**/*.pcap</exclude>
             <exclude>**/*.log1</exclude>
             <exclude>**/*.log2</exclude>
+            <exclude>**/*.sav</exclude>
```
Please also update the exclusion list for `license-maven-plugin`.

cgivre replied:
Done.

paul-rogers commented on contrib/format-spss/README.md:
```diff
@@ -0,0 +1,83 @@
+This format plugin enables Apache Drill to read and query Statistical Package for the Social Sciences (SPSS) (or Statistical Product and Service Solutions) data files. According
+ to Wikipedia: (https://en.wikipedia.org/wiki/SPSS)
+ ***
+ SPSS is a widely used program for statistical analysis in social science. It is also used by market researchers, health researchers, survey companies, government, education researchers, marketing organizations, data miners, and others. The original SPSS manual (Nie, Bent & Hull, 1970) has been described as one of "sociology's most influential books" for allowing ordinary researchers to do their  ...
```
Nit: for ease of editing, it is handy to break lines at around 80 chars. MD will combine them to form a paragraph as if they were one long line.

cgivre replied:
Fixed

paul-rogers commented on contrib/format-spss/README.md:
```diff
@@ -0,0 +1,83 @@
+ ***
+ 
+## Configuration 
+To configure Drill to read SPSS files, simply add the following code to the formats section of your file-based storage plugin.  This should happen automatically for the default
```
Do we want to add all the format plugins at bootstrap? Creates a rather intimidating-looking hunk of JSON for newbies. Of course, it would be good for format plugins to be independent of the storage plugin, but that will come later.

cgivre replied:
My thought is that it's better to add it in the bootstrap so that people know it's there.  Just my .02...

paul-rogers commented on contrib/format-spss/pom.xml:
```diff
@@ -0,0 +1,88 @@
+              <goal>copy-resources</goal>
+            </goals>
+            <configuration>
+              <outputDirectory>${basedir}/target/classes/org/apache/drill/exec/store/syslog
```
`syslog`?

cgivre replied:
Oops ... Fixed

paul-rogers commented on contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java:
```diff
@@ -0,0 +1,240 @@
+  public void close() {
+    if (fsStream != null) {
+      try {
+        fsStream.close();
```
I think `Closeables.closeSilently(fsStream)` is the preferred approach these days.

cgivre replied:
Fixed

paul-rogers commented on contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java:
```diff
@@ -0,0 +1,240 @@
+      throw UserException
+        .dataReadError(e)
+        .message("Error reading SPSS File.")
+        .addContext(e.getMessage())
```
Not necessary. Better is:

```
.context("Error reading SPSS File.")
```
And let the message be the underlying error.

cgivre replied:
I think this is what you meant here.

cgivre replied:
```
throw UserException
        .dataReadError(e)
        .message("Error reading SPSS File.")
        .addContext(errorContext)
        .build(logger);
```

paul-rogers commented on contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java:
```diff
@@ -0,0 +1,240 @@
+  private void buildReaderList() {
+    writerList = new ArrayList<>();
+
+    for(SpssVariable variable : variableList) {
```
Nit: `for (`

cgivre replied:
Fixed

paul-rogers commented on contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java:
```diff
@@ -0,0 +1,240 @@
+  public abstract static class SpssColumnWriter {
+    final String columnName;
+
+    final ScalarWriter writer;
```
Nit: no need to double-space fields. Single-spacing is more compact.

cgivre replied:
Fixed.  I think the Drill auto-formatting did that.

paul-rogers commented on contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java:
```diff
@@ -0,0 +1,240 @@
+      this.writer = writer;
+    }
+
+
```
Nit: extra newlines.

cgivre replied:
Fixed

paul-rogers commented on contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java:
```diff
@@ -0,0 +1,240 @@
+
+    ScalarWriter labelWriter;
+
+    Map<Double, String> labels;
```
This makes me nervous: `double` is a fragile thing to map from. Does SPSS require that indexed columns have integer values? If so, map from an `Integer`, which is more reliable.

cgivre replied:
This is weird, but unfortunately, the only data types this format supports are strings and numerics which are floating point numbers.

paul-rogers commented on contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java:
```diff
@@ -0,0 +1,240 @@
+      if (reader.getValueLabels(columnName) != null && reader.getValueLabels(columnName).size() != 0) {
+        labelWriter = rowWriter.scalar(columnName + VALUE_LABEL);
+        labels = reader.getValueLabels(columnName);
+        hasLabels = true;
```
Nit: `hasLabels` is redundant: can check if `labelWriter` is `null` below.

cgivre replied:
Removed

paul-rogers commented on contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java:
```diff
@@ -0,0 +1,240 @@
+    @Override
+    public void load(SpssDataFileReader reader) {
+
+      double value = reader.getDoubleValue(columnName);
```
Nice use of your `SpssColumnWriter` class to avoid a name lookup for each column write. I wonder, does SPSS provide an indexed way to get values? Do the values form a row (tuple) in addition to a map?

cgivre replied:
It turns out there is!  SPSS provides a way to access the column by index (int) rather than by column name.

cgivre commented:
@paul-rogers 
Thanks for the review.  I believe I addressed all your comments.
